non_semantic_speech_benchmark/distillation/models.py

```diff
@@ -17,42 +17,92 @@
 
 
-def get_keras_model(bottleneck_dimension, output_dimension, alpha=1.0):
-  """Make a model."""
-  audio_tensor = tf.keras.Input((None,))
+def get_keras_model(bottleneck_dimension,
+                    output_dimension,
+                    alpha=1.0,
+                    mobilenet_size='small',
+                    frontend=True,
+                    avg_pool=False):
+  """Make a keras student model."""
   def _map_fn_lambda(x):
     return tf.map_fn(_sample_to_features, x, dtype=tf.float64)
-  feats = tf.keras.layers.Lambda(_map_fn_lambda)(audio_tensor)
-  feats.shape.assert_is_compatible_with([None, None, 96, 64])
-  feats = tf.transpose(feats, [0, 2, 1, 3])
-  feats = tf.reshape(feats, [-1, 96, 64, 1])
-  model = tf.keras.applications.MobileNetV3Large(
+  def _map_mobilenet_func(mnet_size):
+    mnet_size_map = {
+      "tiny": MobileNetV3Tiny,
+      "small": tf.keras.applications.MobileNetV3Small,
+      "large": tf.keras.applications.MobileNetV3Large,
+    }
+    assert mnet_size in mnet_size_map
+    return mnet_size_map[mnet_size.lower()]
+  if frontend:
+    model_in = tf.keras.Input((None,), name="audio_samples")
+    feats = tf.keras.layers.Lambda(_map_fn_lambda)(model_in)
+    feats.shape.assert_is_compatible_with([None, None, 96, 64])
+    feats = tf.transpose(feats, [0, 2, 1, 3])
+    feats = tf.reshape(feats, [-1, 96, 64, 1])
+  else:
+    model_in = tf.keras.Input((96, 64, 1), name="log_mel_spectrogram")
+    feats = model_in
+  model = _map_mobilenet_func(mobilenet_size)(
       input_shape=[96, 64, 1],
       alpha=alpha,
       minimalistic=False,
       include_top=False,
       weights=None,
-      pooling=None,
+      pooling='avg' if avg_pool is True else None,
       dropout_rate=0.0)
-  model_out = model(feats)
-  model_out.shape.assert_is_compatible_with([None, 3, 2, 1280])
-  embeddings = tf.keras.backend.batch_flatten(model_out)
-  embeddings.set_shape([None, 3 * 2 * 1280])
+  if avg_pool:
+    model_out = model(feats)
+    model_out.shape.assert_is_compatible_with([None, None])
+  else:
+    model_out = model(feats)
+    model_out.shape.assert_is_compatible_with([None, 3, 2, None])
+  if bottleneck_dimension:
+    embeddings = tf.keras.layers.Flatten()(model_out)
+    embeddings = tf.keras.layers.Dense(
+      bottleneck_dimension, activation=tf.nn.swish, name='distilled_output')(embeddings)
+  else:
+    embeddings = tf.keras.layers.Flatten(name='distilled_output')(model_out)
   # TODO(joelshor): These final layers can be large. Investigate the compression
   # techniques described in
   # https://blog.tensorflow.org/2020/02/matrix-compression-operator-tensorflow.html?m=1
-  embeddings = tf.keras.layers.Dense(
-      bottleneck_dimension, name='distilled_output')(embeddings)
-  output = tf.keras.layers.Dense(
-      output_dimension, name='embedding_to_target')(embeddings)
+  output = tf.keras.layers.Dense(output_dimension, activation=tf.nn.swish, name='embedding_to_target')(embeddings)
```

Comment:
Why did you add `tf.nn.swish` to the final layer? that seems...wrong, no?

Reply:
Good catch - see my below comment.